File: src/python/Primes.py
def generate(maxValue):
    if maxValue >= 2:
        s = maxValue + 1
        flags = [True] * s
        flags[0] = flags[1] = False
        for i in range(2, int(s**0.5) + 1):
            if flags[i]:
                for j in range(2 * i, s, i):
                    flags[j] = False  # not prime

        # count the number of primes
        count = sum(flags)
        primes = [0] * count
        # move the primes into the result
        j = 0
        for i in range(s):
            if flags[i]:
                primes[j] = i
                j += 1
        return primes
    else:
        return []

import unittest
class TestPrimes(unittest.TestCase):
    def test_up_to_1(self):
        self.assertEqual(generate(1), [])

    def test_up_to5(self):
        self.assertEquals(generate(5), [2, 3, 5])

    def test_up_to10(self):
        self.assertEquals(generate(10), [2,3,5,7])

if __name__ == '__main__':
    unittest.main()

File: src/main/java/com/scrumcenter/csd/refactoring/simple/Primes.java
package com.scrumcenter.csd.refactoring.simple;

public class Primes {
    public static int[] generate(int maxValue) {
        if (maxValue >= 2) {
            int s = maxValue + 1;
            boolean[] flags = new boolean[s];
            int i;
            for (i = 0; i < s; i++)
                flags[i] = true;
            flags[0] = flags[1] = false;
            int j;
            for (i = 2; i < Math.sqrt(s) + 1; i++) {
                if (flags[i]) {
                    for (j = 2 * i; j < s; j += i)
                        flags[j] = false; // not prime
                }
            }

            // count the number of primes
            int count = 0;
            for (i = 0; i < s; i++) {
                if (flags[i])
                    count++; // increase count
            }
            int[] primes = new int[count];
            // move the primes into the result
            for (i = 0, j = 0; i < s; i++) {
                if (flags[i])
                    primes[j++] = i;
            }
            return primes;
        } else
            return new int[0];
    }
}